pilot/pkg/security/authn/v1alpha1/policy_applier.go

```diff
@@ -103,25 +116,84 @@ func outputLocationForJwtIssuer(issuer string) string {
 	return locationPrefix + fmt.Sprintf("%x", sum)
 }
 
-// ConvertPolicyToJwtConfig converts policy into Jwt filter config for envoy.
-func convertPolicyToJwtConfig(policy *authn_v1alpha1.Policy) *jwtfilter.JwtAuthentication {
-	policyJwts := collectJwtSpecs(policy)
-	if len(policyJwts) == 0 {
-		return nil
+func convertToEnvoyJwtConfig(policyJwts []*authn_v1alpha1.Jwt) *envoy_jwt.JwtAuthentication {
+	var requirements []*envoy_jwt.JwtRequirement
+	providers := map[string]*envoy_jwt.JwtProvider{}
+	for _, policyJwt := range policyJwts {
+		provider := &envoy_jwt.JwtProvider{
+			Issuer:               policyJwt.Issuer,
+			Audiences:            policyJwt.Audiences,
+			ForwardPayloadHeader: outputLocationForJwtIssuer(policyJwt.Issuer),
+			Forward:              true,
+			PayloadInMetadata:    policyJwt.Issuer,
+		}
+
+		for _, location := range policyJwt.JwtHeaders {
+			provider.FromHeaders = append(provider.FromHeaders, &envoy_jwt.JwtHeader{
+				Name: location,
+			})
+		}
+		provider.FromParams = policyJwt.JwtParams
+
+		jwtPubKey, err := authn_model.JwtKeyResolver.GetPublicKey(policyJwt.JwksUri)
```

Comment:
If we only try this once, what will happen if we cannot fetch the public key? Shouldn't we leave the remote jwks address for Envoy jwt filter to handle the fetching?

Reply:
we will return empty if we cannot fetch the public key, it will be fetched again in the refresh job or with another config push. I think we decided to move the public key fetch from envoy to pilot long time ago. cc @diemtvu

Reply:
If I understand it correctly, for the public key, Pilot only push a new Envoy config if Pilot gets new resources from Galley or if the refresh job is triggered. So if the jwt policy is the last one, and when Pilot fails to retrieve the public key, the jwt filter won't be able to validate/reject requests until the refresh job is triggered or a new resource is added in. In such case, why don't we just include the remote jwks to the filter for it to fetch the public key? When Pilot is able to get the public key, it can send out the public key instead of the jwks url.

Reply:
Let me know if I'm missing anything @yangminzhu @diemtvu @quanjielin

Reply:
The problem with Envoy is you need to configure a service entry / egress to reach the jwksURI (depends on installation mode, external service is blocked by default). We think have control plane handling this is easier for users (also easier to debug, imo). However, we can revisit this decision with the upstream work.

Reply:
Makes sense to me. Then maybe we should add some note or TODO to prevent us from forgetting this.

Reply:
Nah. We will forget TODO ;). I don't think there is any the problem with this approach yet, and there could be more issues with decentralize model. Let's brainstorm of what we are trying to improve first.